src/epoch/garbage.rs

```diff
@@ -0,0 +1,210 @@
+        //       think, especially if several threads constantly creates new garbage).
+        loop {
+            // Load the head.
+            let head = self.head.load(atomic::Ordering::Acquire);
+            // Store the loaded head in the node which might become the new head node.
+            unsafe { (*n).next.store(head, atomic::Ordering::Relaxed); }
+            // To avoid leaking nodes, we use CAS to ensure that our snapshot (`head`) is still
+            // valid. If it is, we can replace the node with node `n` linking it and break the
+            // loop. If not, we will have to spin again until the ABA condition is over.
+            if self.head.compare_and_swap(head, n, atomic::Ordering::Release) == head {
+                break;
+            }
+        }
+    }
+
+    /// Run the destructor of all the garbage.
+    pub unsafe fn collect(&self) {
+        // Fetch the head node.
+        let mut head = self.head.load(atomic::Ordering::Relaxed);
+        // Make sure it isn't null (if it is, the stack is empty).
+        if head != ptr::null_mut() {
+            // Take the head and leave a null pointer in its place (clear the stack).
+            head = self.head.swap(ptr::null_mut(), atomic::Ordering::Acquire);
+
+            // Now that we have obtained exclusive access to the stack, we can simply iterate over
+            // the nodes and collect one by one.
+            while head != ptr::null_mut() {
+                // Reconstruct the node's pointer.
+                let mut n = Box::from_raw(head);
+                // Then run destructors of the bag.
+                n.data.collect();
+                // The go to next node.
+                head = n.next.load(atomic::Ordering::Relaxed);
+                // Repeat until no more nodes.
+            }
+        }
+    }
+}
+
+// Null pointer is a valid and null `Vec<T>` is also.
+unsafe impl ZerosValid for ConcBag {}
+
+/// A cross-thread set of garbage bags.
+#[derive(Debug, Default)]
+pub struct Global {
+    /// The garbage bags.
+    ///
+    /// This is in order: old, current, new.
+    bags: [ConcBag; 3],
+}
+
+impl Global {
+    /// Create an empty, global set of garbage.
+    // TODO: When stabilized, this can be merged with `Default`.
+    #[cfg(feature = "nightly")]
+    pub const fn new() -> Global {
+        Global {
+            old: ConcBag::zeroed(),
+            cur: ConcBag::zeroed(),
+            new: ConcBag::zeroed(),
+        }
+    }
+
+    /// Collect the garbage of some epoch.
+    pub fn collect(&self, epoch: usize) {
+        // Find the associated array entry and collect it. Due to the invariants of this type, this
+        // is safe.
+        unsafe { self.bags[epoch % 3].collect() }
+    }
+
+    /// Insert a bag of garbage into some epoch..
+    pub fn insert(&self, epoch: usize, bag: Bag) {
+        // Find the associated array entry and insert the bag.
```

Comment:
Instead of stating "find the associated array entry" why not just create a private method called `find_entry`?